From 00ddc42a15182874a5b9619ffd513cfbc5110967 Mon Sep 17 00:00:00 2001
From: Martin Jansa <martin.jansa@gmail.com>
Date: Mon, 28 Sep 2026 21:33:19 +0200
Subject: [PATCH] cert: Fix build with OpenSSL 4.0

OpenSSL 4.0 made ASN1_STRING opaque, so direct access to ->data and
->length fields no longer compiles. Use ASN1_STRING_get0_data() and
ASN1_STRING_length() accessors which are available since OpenSSL 1.1.0.

Upstream-Status: Pending [webOS OSE is no longer maintained on github]
Signed-off-by: Martin Jansa <martin.jansa@gmail.com>
---
 src/cert_db.c   |  8 ++++----
 src/cert_x509.c | 12 ++++++------
 2 files changed, 10 insertions(+), 10 deletions(-)

diff --git a/src/cert_db.c b/src/cert_db.c
index ba8f954..c096e13 100644
--- a/src/cert_db.c
+++ b/src/cert_db.c
@@ -544,8 +544,8 @@ CertReturnCode_t CertCreateDatabaseItem(X509 *x509, char *name, int32_t serial,
           Ansitm = X509_get_notAfter(x509);
 
           row[CERT_DATABASE_ITEM_EXPIRATION] =
-            (char *)newMem(Ansitm->data,
-                           Ansitm->length + 1);
+            (char *)newMem(ASN1_STRING_get0_data(Ansitm),
+                           ASN1_STRING_length(Ansitm) + 1);
           if (NULL == row[CERT_DATABASE_ITEM_EXPIRATION])
             {
               result = CERT_MEMORY_ERROR;
@@ -591,8 +591,8 @@ CertReturnCode_t CertCreateDatabaseItem(X509 *x509, char *name, int32_t serial,
           Ansitm = X509_get_notBefore(x509);
 
           row[CERT_DATABASE_ITEM_START] =
-            (char *)newMem(Ansitm->data,
-                           Ansitm->length + 1);
+            (char *)newMem(ASN1_STRING_get0_data(Ansitm),
+                           ASN1_STRING_length(Ansitm) + 1);
           if (NULL == row[CERT_DATABASE_ITEM_START])
             {
               result = CERT_MEMORY_ERROR;
diff --git a/src/cert_x509.c b/src/cert_x509.c
index 0e7e5d2..a0ea522 100644
--- a/src/cert_x509.c
+++ b/src/cert_x509.c
@@ -111,9 +111,9 @@ int CertX509ReadStrProperty(X509 *cert, int property, char *pBuf, int len)
 			/* Do something with e */
 			ASN1_IA5STRING *data;
 			data = X509_NAME_ENTRY_get_data(e);
-			syslog(LOG_INFO,"all common name: %s", (char *) data->data);
+			syslog(LOG_INFO,"all common name: %s", (char *) ASN1_STRING_get0_data(data));
 			if (0 < space_left) {
-				space_taken = copy_csv_to_buffer(sub_str, (char *)data->data, len, space_left);
+				space_taken = copy_csv_to_buffer(sub_str, (char *)ASN1_STRING_get0_data(data), len, space_left);
 
 				/* 0 means the value did not fit and was truncated */
 				if (0 == space_taken)
@@ -168,7 +168,7 @@ int get_subjectaltname(X509* cert, char* buf, int buf_len){
 
       if((gen->type == GEN_DNS)||(gen->type == GEN_URI)){
           if(0 < space_left)
-	  space_taken= copy_csv_to_buffer(sub_str, (char*)gen->d.ia5->data, buf_len, space_left);
+	  space_taken= copy_csv_to_buffer(sub_str, (char*)ASN1_STRING_get0_data(gen->d.ia5), buf_len, space_left);
 	  space_left= space_left -space_taken;
       }
 
@@ -199,12 +199,12 @@ int ip_to_string(char* oline, int oline_len, GENERAL_NAME* gen)
   unsigned char *p;
   char htmp[5];
 
-  p = gen->d.ip->data;
+  p = ASN1_STRING_get0_data(gen->d.ip);
 
-  if(gen->d.ip->length == 4)
+  if(ASN1_STRING_length(gen->d.ip) == 4)
     BIO_snprintf(oline, oline_len,"%d.%d.%d.%d", p[0], p[1], p[2], p[3]);  //sizeof oline replaced by 40
 
-  else if(gen->d.ip->length == 16){
+  else if(ASN1_STRING_length(gen->d.ip) == 16){
     oline[0] = 0;
     for (i = 0; i < 8; i++){
       BIO_snprintf(htmp, sizeof htmp,"%X", p[0] << 8 | p[1]);
